### PR TITLE
Pa02

### PR DESCRIPTION
**Pull request by elinwub**
Handing in for Aleksander Eriksen and Elin Bjørnson

### CODE REVIEW

File: src/pa02/chutes_simulation.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+            position where the goal is reached
+        """
+
+        if ladders and chutes and goal is not None:
+            self.board = {**dict(ladders), **dict(chutes)}
+            self.goal = goal
+
+        else:
+            self.board = Board.std_board
+            self.goal = Board.std_goal
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:35]: Smart, but when one of `ladders` or `snakes` is `None`, then the default board is used?

File: src/pa02/chutes_simulation.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    def __init__(self, board=Board()):
+        """
+        Initialise Player
+        Parameters
+        ----------
+        board: class
+            the board used in the game.
+        """
+        self.board = board
+        self.position = 0
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:77]: Need number of steps as attribute, which shouldn't be calculate outside this class.

File: src/pa02/chutes_simulation.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            List of players in game
+        board: class
+            Board used in game
+        seed: int
+        randomize_players: bool
+            option to randomize order of players
+        """
+        self.players = player_field
+        self.board = board
+        self.seed = random.seed(seed)
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:181]: No need storing seed here.

File: src/pa02/chutes_simulation.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        self.randomize_players = randomize_players
+        self.turns = 0
+        self.results = []
+        if self.randomize_players is True:
+            random.shuffle(self.players)
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:186]: It should be done in every single game. Not only once in here.

File: src/pa02/chutes_simulation.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+        """
+        Runs single game, returns number of moves and the type of player who
+        won
+        Returns
+        -------
+        tuple (num_moves, 'player_type')
+        """
+
+        players = [p(self.board) for p in self.players]
+        self.turns = 0
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:198]: Check comment in player about number of steps.

File: src/pa02/chutes_simulation.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    def durations_per_type(self):
+        """
+        Returns dictionary mapping player types to lists of game durations
+        of each type of player
+        Returns
+        -------
+        dict
+        """
+
+        durations = defaultdict(list)
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:254]: Nice thinking.

File: src/pa02/test_pa02.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+
+class TestBoard:
+    def test_only_one_input(self):
+        """
+        Tests if we get get standard board if we only have one input
+        """
+        b_std = cs.Board()
+        b = cs.Board([(1, 4), (5, 16)])
+        assert b.board == b_std.board
```
Comment (huynhngoc) [inline src/pa02/test_pa02.py:17]: Okay, this test supports the logic in your constructor in `Board`. Not wrong

File: src/pa02/test_pa02.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+
+class TestPlayerTypes:
+    def test_player_position_changes(self):
+        """
+        Tests that player position changes after a move
+        """
+        p = cs.Player()
+        p.move()
+        assert p.position != 0
```
Comment (huynhngoc) [inline src/pa02/test_pa02.py:71]: `>` would be better.

File: src/pa02/test_pa02.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        p = [cs.Player, cs.Player, cs.ResilientPlayer, cs.ResilientPlayer,
+                          cs.ResilientPlayer, cs.LazyPlayer]
+        sim1 = cs.Simulation(p, randomize_players=True)
+        sim2 = cs.Simulation(p, randomize_players=False)
+
+        sim1.run_simulation(10)
+        sim2.run_simulation(10)
+
+        assert sim1.get_results() is not sim2.get_results()
```
Comment (huynhngoc) [inline src/pa02/test_pa02.py:128]: 👍

**Review by huynhngoc — APPROVED** [review]
Nice try. Approved.

Please check the comments.

File: src/pa02/chutes_simulation.py (hunk 1 above)
Comment (huynhngoc) [reply src/pa02/chutes_simulation.py:35]: Saw that it's how you understand the logic in your test case.

**Review by jakeriksen — APPROVED** [review]

**Pull request merged**